### ISSUE TITLE
Changed behavior of deployment outputs variable

### ISSUE THREAD

**Issue by gbrueckl** [microsoft/azure-pipelines-tasks#14097]
## Note  
The way how the deployment outputs variable of the **Azure resource group deployment** task seem to have changed from task version 2.175.2 to 2.179.0

## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: here

**Enter Task Name**: here  

list here (V# not needed):  
https://github.com/Microsoft/azure-pipelines-tasks/tree/master/Tasks

## Environment
Azure DevOps Pipelines:
actually I am not allowed to post the required details here 

Agent:
Azure Pipelines
windows-2019
## Issue Description
in 2.175.2 it used to create a single output only: 
```
2020-11-02T08:29:44.8029406Z Deployment name is azuredeploy-20201102-082941-cc15
2020-11-02T08:32:22.7443770Z Updated output variable 'armOutput', which contains the outputs section of the current deployment object in string format.
2020-11-02T08:32:22.7445271Z Successfully deployed the template.
```

in 2.179.0 it seems to set multiple variables:
```
2020-12-17T16:26:26.5030124Z Deployment name is azuredeploy-20201217-162624-fd13
2020-12-17T16:28:28.1914050Z Updated output variable 'armOutput.resourceGroupName.type', which contains the outputs section of the current deployment object in string format.
2020-12-17T16:28:28.1919956Z Updated output variable 'armOutput.resourceGroupName.value', which contains the outputs section of the current deployment object in string format.
2020-12-17T16:28:28.1924211Z Updated output variable 'armOutput.resourceGroupId.type', which contains the outputs section of the current deployment object in string format.
2020-12-17T16:28:28.1935561Z Updated output variable 'armOutput.resourceGroupId.value', which contains the outputs section of the current deployment object in string format.
....
```
I used to read the output variable in the following PowerShell task using
`$armJson = $env:armOutput | ConvertFrom-Json`

this fails now as the environment variable "armOutput" itself does not exist anymore

The "outputs" in my ARM template are defined as:
```
  "outputs": {
    "resourceGroupName": {
      "value": "[variables('resourceGroupName')]",
      "type": "string"
    },
    "resourceGroupId": {
      "value": "[resourceGroup().id]",
      "type": "string"
    },
    "adminUser": {
      "value": "[parameters('adminUser')]",
      "type": "string"
    },
    "deployedResources": {
      "type": "object",
      "value": "[reference(variables('deploymentName')).outputs.deployedResources.value]"
    }
  }
```

### Task logs
They would contain confidential information

### Error logs

I do get the error when accessing the output in the following PowerShell task using 
`$armJson = $env:armOutput | ConvertFrom-Json`
```
2020-12-17T16:28:39.4473717Z ##[error]Cannot bind argument to parameter 'InputObject' because it is null.
2020-12-17T16:28:39.5582902Z ##[error]PowerShell exited with code '1'.
```

**Comment by gbrueckl** [comment]
it further seems like not all outputs are generated as individual environment variables!

**Comment by Viper61x23** [comment]
> it further seems like not all outputs are generated as individual environment variables!

@gbrueckl Hi did you manage to get past this issue, I have exactly the same thing as you reference here but I haven't found a way to address, or if its a bug or not.

**Comment by gbrueckl** [comment]
hi @Viper61x23
actually I am quite sure it is a bug - actually two bugs
1) the behavior of how the output variable is used was completely changed from version 2.175.2 to 2.179.0
2) 2.179.0 does not output all values that are defined in the ARM template

to at least work around issue 1) I wrote a PowerShell script that parses the single environment variables and puts them back into one single PowerShell object:
my output-variable as defined in the DevOps task is called "armOutput"
```
	$ARMOutput = $env:armOutput

	# there is a bug with the output of the Deploy ARM template task in the CI/CD pipeline
	# instead of creating a single environment variable holding the whole JSON, single variables
	# are created which need to be merged into the full object manually!
	if(-not $ARMOutput)
	{
		Write-Warning "Parsing ARM Output from environment variables!"
		Get-ChildItem -Path Env:\ | ForEach-Object { Write-Verbose "$($_.Name)    -    $($_.Value)"}

		$envVars = Get-ChildItem -Path Env:\ | Where-Object { $_.Name.ToUpper().startsWith('ARMOUTPUT_')}

		$envVarObj = @{}
		$arrayObjectsAdded = @()

		foreach($envVar in $envVars)
		{
			Write-Information "Processing Environment variable '$($envVar.Name)' ..."
			$tokens = $envVar.Name.Split('_')
			$obj = $envVarObj

			for ($i = 1; $i -lt $tokens.Count; $i++)
			{
				$token = $tokens[$i]
				if($token -match '^[0-9]+$')
				{
					if($arrayObjectsAdded -notcontains $token)
					{
						$obj.Add(@{})
						$arrayObjectsAdded += $token
					}
					$obj = $obj[$arrayObjectsAdded.Count-1]
				}
				elseif(-not $obj.Contains($token))
				{
					$obj.Item($token) = @{}
					if($tokens[$i+1] -match '^[0-9]+$')
					{
						$obj.Item($token) = New-Object System.Collections.Generic.List[System.Object]
						$arrayObjectsAdded = @()
					}
					if($i -eq  $tokens.Count -1)
					{
						$obj.Item($token) = $envVar.Value.Trim('"')
					}
					$obj = $obj.Item($token)
				}
				else {
					$obj = $obj.Item($token)
				}
			}
		}

		$ARMOutput = $envVarObj | ConvertTo-Json -Depth 15
	}

	$armJson = $ARMOutput | ConvertFrom-Json
```

please note, my output contains complex nested objects and arrays, thats why the script is so complex. if you only have scalar values, it can probably be simplified a lot

**Comment by ddivanshu** [comment]
Though in the updated version, task tends to flatten the variable but that comes as additional variable without impacting the current variable.Although there was an issue https://github.com/microsoft/azure-pipelines-tasks/issues/14246 that if one of the field of deployment output was null, output variable was not being populated which has been fixed in the version 2.182.
This could be the reason for your issue as well.Could you try running the task again without the custom power-shell script.

**Comment by ddivanshu** [comment]
Closing this as issue is resolved with updated version.
Feel free to reopen if issue persists.
